### PR TITLE
GG-33208 Earliest checkpoints map persistence for a faster node restart

### PR DESCRIPTION
**Pull request by SammyVimes**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java [hunk 1]
```diff
@@ -108,13 +110,30 @@
...
+                CheckpointEntry entry = entry(timestamp);
+
+                UUID checkpointId = entry.checkpointId();
+
+                Map<Integer, CheckpointEntry.GroupState> groupStateMap = snapshot.getGroupState(checkpointId);
+
+                // Ignore checkpoint that was present at the time of the snapshot and whose group
+                // states map was not persisted (that means this checkpoint wasn't a part of earliestCp map)
+                if (snapshot.checkpointWasPresent(checkpointId) && groupStateMap == null)
+                    continue;
```
Comment (vldpyatkov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java:131]: Can we go deeper in the history when the earlier checkpoint was skipped?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointManager.java [hunk 2]
```diff
@@ -136,15 +136,17 @@ public CheckpointManager(
 
         FileIOFactory ioFactory = persistenceCfg.getFileIOFactory();
 
+        CheckpointReadWriteLock lock = new CheckpointReadWriteLock(logger);
```
Comment (vldpyatkov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointManager.java:139]: Is this lock created only to pass it to method?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointMarkersStorage.java [hunk 3]
```diff
@@ -535,4 +608,100 @@ public static String checkpointFileName(CheckpointEntry cp, CheckpointEntryType
...
+     */
+    void onEarliestCheckpointMapChanged() {
+        boolean createCheckpoint =
+            checkpointSnapshotCounter.getAndUpdate(old -> (old + 1) % EARLIEST_CP_CHANGES_THRESHOLD) == 0;
+
+        if (createCheckpoint) {
+            Runnable runnable = () -> {
+                EarliestCheckpointMapSnapshot snapshot;
+
+                lock.readLock();
```
Comment (vldpyatkov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointMarkersStorage.java:638]: It is a suspicious lock.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointMarkersStorage.java [hunk 4]
```diff
@@ -535,4 +608,100 @@ public static String checkpointFileName(CheckpointEntry cp, CheckpointEntryType
...
+    /**
+     * Handles changes in the earliest checkpoint map in the {@link CheckpointHistory}. Creates a snapshot of that
+     * map if threshold for changes has been reached.
+     */
+    void onEarliestCheckpointMapChanged() {
+        boolean createCheckpoint =
+            checkpointSnapshotCounter.getAndUpdate(old -> (old + 1) % EARLIEST_CP_CHANGES_THRESHOLD) == 0;
+
+        if (createCheckpoint) {
+            Runnable runnable = () -> {
```
Comment (vldpyatkov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointMarkersStorage.java:635]: Think about using a common way with system thread pool and timeout worker.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointManager.java (hunk 2 above)
Comment (SammyVimes) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointManager.java:143]: No it's a checkpoint lock, it's also used by checkpointer

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointMarkersStorage.java (hunk 3 above)
Comment (SammyVimes) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointMarkersStorage.java]: Well I have to lock on the checkpoint lock so I can make a snapshot of checkpoint map

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java [hunk 5]
```diff
@@ -106,15 +108,33 @@
     }
 
     /**
+     * @param snapshot Earliest checkpoint map snapshot.
      * @param checkpoints Checkpoints.
      */
-    public void initialize(List<CheckpointEntry> checkpoints) {
+    public void initialize(
+        EarliestCheckpointMapSnapshot snapshot,
```
Comment (tkalkirill) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java:117]: It is customary to declare new fields and methods below. WDYT?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java [hunk 6]
```diff
@@ -196,11 +216,16 @@ public void addCheckpoint(CheckpointEntry entry, Map<Integer, CacheState> cacheS
     /**
      * Update map which stored the earliest checkpoint each partitions for groups.
      *
+     * @param groupStateMapFromSnapshot Group state map from the snapshot.
      * @param entry Checkpoint entry.
      */
-    private void updateEarliestCpMap(CheckpointEntry entry) {
+    private void updateEarliestCpMap(
+        @Nullable Map<Integer, CheckpointEntry.GroupState> groupStateMapFromSnapshot,
```
Comment (tkalkirill) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java]: It is customary to declare new fields and methods below. WDYT?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointMarkersStorage.java [hunk 7]
```diff
@@ -137,7 +174,34 @@ public void cleanupCheckpointDirectory() throws IgniteCheckedException {
      * Filling internal structures with data from disk.
      */
     public void initialize() throws IgniteCheckedException {
-        cpHistory.initialize(retrieveHistory());
+        File snapshotFile = new File(cpDir, EARLIEST_CP_SNAPSHOT_FILE);
+        File snapshotTmpFile = new File(cpDir, EARLIEST_CP_SNAPSHOT_TMP_FILE);
+
+        if (snapshotTmpFile.exists()) {
+            if (!IgniteUtils.delete(snapshotTmpFile))
+                log.error("Failed to remove invalid earliest checkpoint map snapshot temporary file: " + snapshotFile);
```
Comment (tkalkirill) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointMarkersStorage.java:182]: Can throw an exception and ask to delete the file by hand?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointMarkersStorage.java [hunk 8]
```diff
@@ -137,7 +174,34 @@ public void cleanupCheckpointDirectory() throws IgniteCheckedException {
...
+            try {
+                byte[] bytes = Files.readAllBytes(snapshotFile.toPath());
+
+                snap = JdkMarshaller.DEFAULT.unmarshal(bytes, null);
+            }
+            catch (IOException e) {
+                log.error("Failed to unmarshal earliest checkpoint map snapshot", e);
+
+                if (!IgniteUtils.delete(snapshotFile))
+                    log.error("Failed to remove invalid earliest checkpoint map snapshot file: " + snapshotFile);
```
Comment (tkalkirill) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointMarkersStorage.java:197]: Can throw an exception and ask to delete the file by hand?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointMarkersStorage.java [hunk 9]
```diff
@@ -65,28 +88,41 @@
...
+    private final CheckpointReadWriteLock lock;
+
     /** Checkpoint metadata directory ("cp"), contains files with checkpoint start and end */
     public final File cpDir;
 
     /** Temporary write buffer. */
     private final ByteBuffer tmpWriteBuf;
 
+    /** Counter representing quantity of checkpoints since last checkpoint history snapshot. */
+    private final AtomicInteger checkpointSnapshotCounter = new AtomicInteger(1);
```
Comment (tkalkirill) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointMarkersStorage.java:101]: I think it's better **AtomicLong** or use **AtomicInteger#updateAndGet** with reset value

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/EarliestCheckpointMapSnapshot.java [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+import java.util.Set;
+import java.util.UUID;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Earliest checkpoint map snapshot.
+ * Speeds up construction of the earliestCp map in the {@link CheckpointHistory}.
+ */
+public class EarliestCheckpointMapSnapshot implements Externalizable {
```
Comment (tkalkirill) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/EarliestCheckpointMapSnapshot.java]: Please use **IgniteDataTransferObject**

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/EarliestCheckpointMapSnapshot.java [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+        checkpointIds = new HashSet<>();
+    }
+
+    /**
+     * Gets a group state by a checkpoint id.
+     *
+     * @param checkpointId Checkpoint id.
+     * @return Group state.
+     */
+    @Nullable
```
Comment (tkalkirill) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/EarliestCheckpointMapSnapshot.java:62]: Please move to the line below

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsCheckpointMapSnapshotTest.java [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+
+        configuration.setDataStorageConfiguration(
+            new DataStorageConfiguration().setDefaultDataRegionConfiguration(
+                new DataRegionConfiguration().setPersistenceEnabled(true)
+            ).setCheckpointFrequency(TimeUnit.HOURS.toMillis(1))
+        );
+
+        // Plugin that creates a WAL manager that counts replays
+        configuration.setPluginProviders(new AbstractTestPluginProvider() {
+            @Override public String name() {
```
Comment (tkalkirill) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsCheckpointMapSnapshotTest.java:70]: /** {@inheritDoc} */

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsCheckpointMapSnapshotTest.java [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+                return "testPlugin";
+            }
+
+            @Override public <T> @Nullable T createComponent(PluginContext ctx, Class<T> cls) {
```
Comment (tkalkirill) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsCheckpointMapSnapshotTest.java:74]: /** {@inheritDoc} */

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsCheckpointMapSnapshotTest.java [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+    private static class TestFileWriteAheadLogManager extends FileWriteAheadLogManager {
+        /** Count of times that {@link #replay(WALPointer, IgniteBiPredicate)} has been called. */
+        private final AtomicInteger replayCount = new AtomicInteger();
+
+        /** Constructor. */
+        public TestFileWriteAheadLogManager(GridKernalContext ctx) {
+            super(ctx);
+        }
+
+        private static final String clsName = GroupStateLazyStore.class.getName();
```
Comment (tkalkirill) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsCheckpointMapSnapshotTest.java:95]: javadoc

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsCheckpointMapSnapshotTest.java [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+
+            return super.replay(start, recordDeserializeFilter);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        super.beforeTest();
+
+        cleanPersistenceDir();
```
Comment (tkalkirill) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsCheckpointMapSnapshotTest.java:116]: add stopAllGrids()

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsCheckpointMapSnapshotTest.java [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+     */
+    private void testRestart(boolean removeSnapshot) throws Exception {
+        IgniteEx grid = startGrid(0);
+
+        grid.cluster().state(ClusterState.ACTIVE);
+
+        // Count of inserts and checkpoints
+        int cnt = 100;
+
+        CacheConfiguration<Integer, Integer> configuration = new CacheConfiguration<>("some-cache");
```
Comment (tkalkirill) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsCheckpointMapSnapshotTest.java:162]: Use **GridAbstractTest#DEFAULT_CACHE_NAME**

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsCheckpointMapSnapshotTest.java [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+        IgniteCache<Integer, Integer> cache = grid.getOrCreateCache(configuration);
+
+        for (int i = 0; i < cnt; i++) {
+            cache.put(i, i);
+
+            forceCheckpoint(grid);
+        }
+
+        // Sleep enough to trigger checkpoint map snapshot capture
+        U.sleep(300);
```
Comment (tkalkirill) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsCheckpointMapSnapshotTest.java:173]: Flaky way

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsCheckpointMapSnapshotTest.java [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+
+        stopGrid(0, true);
+
+        if (removeSnapshot) {
+            // Remove checkpoint map snapshot
+            File cpDir = dbMgr(grid).checkpointManager.checkpointDirectory();
+
+            File cpSnapshotMap = new File(cpDir, CheckpointMarkersStorage.EARLIEST_CP_SNAPSHOT_FILE);
+
+            IgniteUtils.delete(cpSnapshotMap);
```
Comment (tkalkirill) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsCheckpointMapSnapshotTest.java:183]: assertTrue i guess

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointMarkersStorage.java (hunk 9 above)
Comment (SammyVimes) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointMarkersStorage.java:101]: Why `long` if `earliestCpChangesThreshold` is `int`?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointEntry.java [hunk 19]
```diff
@@ -269,27 +314,26 @@ private GroupStateLazyStore() {
...
 
         /**
          * @param stateRec Cache group state.
          */
-        private Map<Integer, GroupState> remap(Map<Integer, CacheState> stateRec) {
-            if (stateRec == null)
-                return Collections.emptyMap();
-
+        private static Map<Integer, GroupState> remap(@NotNull Map<Integer, CacheState> stateRec) {
             Map<Integer, GroupState> grpStates = U.newHashMap(stateRec.size());
```
Comment (tkalkirill) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointEntry.java:337]:
```
if (stateRec.isEmpty()) 
    return emptyMap();
```

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/EarliestCheckpointMapSnapshot.java [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void readExternalData(byte protoVer, ObjectInput in) throws IOException, ClassNotFoundException {
+        data = U.readMap(in);
+        checkpointIds = U.readSet(in);
+    }
+
+    /** {@inheritDoc} */
+    @Override public byte getProtocolVersion() {
```
Comment (tkalkirill) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/EarliestCheckpointMapSnapshot.java:107]: This is the default, I think it should be removed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/EarliestCheckpointMapSnapshot.java [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+        }
+
+        /** {@inheritDoc} */
+        @Override protected void readExternalData(byte protoVer, ObjectInput in) throws IOException, ClassNotFoundException {
+            parts = U.readIntArray(in);
+            cnts = U.readLongArray(in);
+        }
+
+        /** {@inheritDoc} */
+        @Override public byte getProtocolVersion() {
```
Comment (tkalkirill) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/EarliestCheckpointMapSnapshot.java:164]: This is the default, I think it should be removed.

File: modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java [hunk 22]
```diff
@@ -5609,7 +5609,7 @@ public static IgniteUuid bytesToIgniteUuid(byte[] in, int off) {
      * @param arr Array to write, possibly <tt>null</tt>.
      * @throws IOException If write failed.
      */
-    public static void writeBooleanArray(DataOutput out, @Nullable boolean[] arr) throws IOException {
+    public static void writeBooleanArray(DataOutput out, boolean @Nullable [] arr) throws IOException {
```
Comment (tkalkirill) [inline modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java:5612]: Let's not change

File: modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java [hunk 23]
```diff
@@ -5627,7 +5627,7 @@ public static void writeBooleanArray(DataOutput out, @Nullable boolean[] arr) th
      * @param arr Array to write, possibly <tt>null</tt>.
      * @throws IOException If write failed.
      */
-    public static void writeIntArray(DataOutput out, @Nullable int[] arr) throws IOException {
+    public static void writeIntArray(DataOutput out, int @Nullable [] arr) throws IOException {
```
Comment (tkalkirill) [inline modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java:5630]: Let's not change

File: modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java [hunk 24]
```diff
@@ -5638,14 +5638,32 @@ public static void writeIntArray(DataOutput out, @Nullable int[] arr) throws IOE
...
+
     /**
      * Reads boolean array from input stream accounting for <tt>null</tt> values.
      *
      * @param in Stream to read from.
      * @return Read byte array, possibly <tt>null</tt>.
      * @throws IOException If read failed.
      */
-    @Nullable public static boolean[] readBooleanArray(DataInput in) throws IOException {
+    public static boolean @Nullable [] readBooleanArray(DataInput in) throws IOException {
```
Comment (tkalkirill) [inline modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java:5666]: Let's not change

File: modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java [hunk 25]
```diff
@@ -5666,7 +5684,7 @@ public static void writeIntArray(DataOutput out, @Nullable int[] arr) throws IOE
      * @return Read byte array, possibly <tt>null</tt>.
      * @throws IOException If read failed.
      */
-    @Nullable public static int[] readIntArray(DataInput in) throws IOException {
+    public static int @Nullable [] readIntArray(DataInput in) throws IOException {
```
Comment (tkalkirill) [inline modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java:5687]: Let's not change

File: modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java [hunk 26]
```diff
@@ -5680,6 +5698,27 @@ public static void writeIntArray(DataOutput out, @Nullable int[] arr) throws IOE
...
     }
 
+    /**
+     * Reads long array from input stream.
+     *
+     * @param in Stream to read from.
+     * @return Read long array, possibly <tt>null</tt>.
+     * @throws IOException If read failed.
+     */
+    public static long @Nullable [] readLongArray(DataInput in) throws IOException {
```
Comment (tkalkirill) [inline modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java:5708]: Let's write: `long @Nullable []` -> `@Nullable long []`

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsCheckpointMapSnapshotTest.java [hunk 27]
```diff
@@ -0,0 +1,221 @@
...
+
+            assertTrue(!cpSnapshotMap.exists());
```
Comment (tkalkirill) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsCheckpointMapSnapshotTest.java:188]: assertFalse =)

**Pull request merged**